File: pset6/credit/credit.py
from cs50 import get_int

MIN = 4000000000000
MAX = 5600000000000000


def main():
    card = get_int("Number: ")

    if(validate(card)):
        prefix = int(str(card)[0] + str(card)[1])

        if prefix in [34, 37]:
            print("AMEX")
        elif prefix in range(40,50):
            print("VISA")
        elif prefix in range(51,56):
            print("MASTERCARD")
        else:
            print("INVALID")


def validate(card):
    if card < MIN or card > MAX:
        # print("out of bounds")
        return False

    card = str(card)
    digits = []

    for i in range(len(card), 16):
        digits.append(0)

    for d in card:
        digits.append(int(d))

    formula = 0
    for i in range(0, len(digits)):
        formula += digits[i]

        if i % 2 == 0:
            formula += digits[i]

            if digits[i] > 4:
                formula -= 9

    # print(formula)
    if formula % 10 == 0:
        return True
    else:
        return False


if __name__ == "__main__":
    main()

File: pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

#define MIN 4000000000000
#define MAX 5600000000000000

int validate(long card);

int main(void)
{
    // Prompts user for a card number
    long cardNum = get_long("Number: ");

    int value = validate(cardNum);

    // Prints depending on the returned value.
    if (value / 10 == 4)
    {
        printf("VISA\n");
    }
    else if (value > 50 && value < 56)
    {
        printf("MASTERCARD\n");
    }
    else if (value == 34 || value == 37)
    {
        printf("AMEX\n");
    }
    else
    {
        printf("INVALID\n");
    }

}

// Process to validate cards, returns the first two card digits if valid
int validate(long card)
{
    // Detects if the number is outside the acceptable card range
    if (card < MIN || card > MAX)
    {
        return 1;
    }

    int len = 16, num = 0;

    // Converts card number to an array to access each digit
    int cards[len];
    for (int i = len - 1; i >= 0; i--)
    {
        cards[i] = card % 10;
        card /= 10;
    }

    //
    for (int i = 0; i < len; i++)
    {
        if (i % 2 == 1)
        {
            num += cards[i];
        }

        else
        {
            num += cards[i] * 2;

            if (cards[i] > 4)
            {
                num -= 9;
            }
        }

    }

    // Decides to return the first two card digits if the card is valid
    if (num % 10 == 0)
    {
        for (int i = 0; i < len; i++)
        {
            if (cards[i] != 0)
            {
                num = cards[i]*10 + cards[i+1];
                i = len;
            }
        }
    }
    else
    {
        num = 1;
    }

    return num;
}